### PR TITLE
QuarkusUnitTest - detect usage of non-static fields and fail correctly

### PR DESCRIPTION
**Pull request by mkouba**
- if a non-static field is used the test app is not started
- also update docs

### CODE REVIEW

**Review by manovotn — APPROVED** [review]

File: test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java [hunk 1]
```diff
@@ -381,6 +386,20 @@ public void afterEach(ExtensionContext context) throws Exception {
 
     @Override
     public void beforeEach(ExtensionContext context) throws Exception {
```
Comment (geoand) [inline test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java]: Does this need to be done for each test method? Wouldn't it make sense to do it in `beforeAll`?

Comment (mkouba) [reply test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java]: `beforeAll` is not executed at all when a non-static field is used... that's the whole point of the issue ;-)

Comment (geoand) [reply test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java]: OK, thanks

Comment (geoand) [reply test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java]: Although I would have hoped there would be a more useful lifecycle method. Perhaps where the test class is constructed?

Comment (mkouba) [reply test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java]: `TestInstanceFactory.createTestInstance()` is not executed either.

Comment (geoand) [reply test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java]: I rest my case then :)

Comment (manovotn) [reply test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java]: I like how @geoand asks the exact same questions I did xD
BTW this bit documents what is/isn't invoked if the registration is non-static https://junit.org/junit5/docs/current/user-guide/#extensions-registration-programmatic-instance-fields

Comment (geoand) [reply test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java]: Great reference, thanks @manovotn !

**Comment by geoand** [comment]
It seems like some tests in `Arc Deployment` are failing

**Comment by mkouba** [comment]
All the tests that expect a deployment failure seem to fail. Which is weird because for such tests I would expect `beforeEach()` not to be called (the test methods are not invoked). @manovotn any idea?

**Comment by manovotn** [comment]
FTR after a short Zulip exchange we found it was a missing check for tests that assert on deployment failure.

**Pull request merged**